src/app/clusters/general-commissioning-server/general-commissioning-server.cpp

```diff
@@ -57,92 +58,106 @@ using Transport::Session;
-    GeneralCommissioningAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), GeneralCommissioning::Id) {}
-
-    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
-
-private:
-    CHIP_ERROR ReadIfSupported(CHIP_ERROR (ConfigurationManager::*getter)(uint8_t &), AttributeValueEncoder & aEncoder);
-    CHIP_ERROR ReadBasicCommissioningInfo(AttributeValueEncoder & aEncoder);
-    CHIP_ERROR ReadSupportsConcurrentConnection(AttributeValueEncoder & aEncoder);
-};
-
-GeneralCommissioningAttrAccess gAttrAccess;
-
-CHIP_ERROR GeneralCommissioningAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
-{
-    if (aPath.mClusterId != GeneralCommissioning::Id)
+    if (nullptr == provider)
     {
-        // We shouldn't have been called at all.
-        return CHIP_ERROR_INVALID_ARGUMENT;
+        return CHIP_ERROR_PERSISTED_STORAGE_FAILED;
     }
 
-    switch (aPath.mAttributeId)
+    T value;
+    CHIP_ERROR err = (provider->*nonConstGetter)(value);
+    if (err != CHIP_NO_ERROR)
     {
-    case RegulatoryConfig::Id: {
-        return ReadIfSupported(&ConfigurationManager::GetRegulatoryLocation, aEncoder);
-    }
-    case LocationCapability::Id: {
-        return ReadIfSupported(&ConfigurationManager::GetLocationCapability, aEncoder);
-    }
-    case BasicCommissioningInfo::Id: {
-        return ReadBasicCommissioningInfo(aEncoder);
-    }
-    case SupportsConcurrentConnection::Id: {
-        return ReadSupportsConcurrentConnection(aEncoder);
-    }
-    default: {
-        break;
-    }
+        return err;
     }
-    return CHIP_NO_ERROR;
+
+    return aEncoder.Encode(value);
 }
 
-CHIP_ERROR GeneralCommissioningAttrAccess::ReadIfSupported(CHIP_ERROR (ConfigurationManager::*getter)(uint8_t &),
-                                                           AttributeValueEncoder & aEncoder)
+template <typename Provider, typename T>
+static CHIP_ERROR ReadIfSupported(Provider * const provider, CHIP_ERROR (Provider::*const getter)(T &) const,
+                                  AttributeValueEncoder & aEncoder)
 {
-    uint8_t data;
-    CHIP_ERROR err = (DeviceLayer::ConfigurationMgr().*getter)(data);
-    if (err == CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE)
-    {
-        data = 0;
-    }
-    else if (err != CHIP_NO_ERROR)
-    {
-        return err;
-    }
-
-    return aEncoder.Encode(data);
+    // Removing the const qualifier from the getter function pointer because there are a handful of getter functions that are not
+    // correctly marked as const.
+    using NonConstGetter          = CHIP_ERROR (Provider::*)(T &);
+    NonConstGetter nonConstGetter = reinterpret_cast<NonConstGetter>(getter);
+    return ReadIfSupported(provider, nonConstGetter, aEncoder);
 }
 
-CHIP_ERROR GeneralCommissioningAttrAccess::ReadBasicCommissioningInfo(AttributeValueEncoder & aEncoder)
+class _ : public AttributeAccessInterface
```

Comment:
This pattern makes debugging harder as the members become opaque as `_::xxxx`.

Please just use a usuall named class definition, then instantiate the singleton below.